File: testboard-bp/testboard.py
from pyBusPirateLite.I2Chigh import *
from pyBusPirateLite.RAW_WIRE import *
from optparse import OptionParser, OptionGroup
import sys

#--------------------------------------------------------------------
#Functions
#--------------------------------------------------------------------

def setDACValue(device, addr, value):
  #Assemble the 2 data bytes
  datahigh = (value >> 6) & 0xFF
  datalow = (value & 0xFF)

  #Send
  bp = I2Chigh(device, 115200, 1)
  bp.BBmode()
  bp.enter_I2C()
  bp.cfg_pins(I2CPins.PULLUPS)

  bp.send_start_bit()
  stat = bp.bulk_trans(1, [addr<<1 | 0])
  print "stat: %u %x %u" % (stat[0],stat[1],stat[2])
  if stat[0] == chr(0x01):
    raise IOError, "DAC didn't ack address and write"
  stat = bp.bulk_trans(1, [datahigh])
  print "stat: %u %x %u" % (stat[0],stat[1],stat[2])
  if stat[0] == chr(0x01):
    raise IOError, "DAC didn't ack top bits"
  stat = bp.bulk_trans(1, [datalow])
  print "stat: %u %x %u" % (stat[0],stat[1],stat[2])
  if stat[0] == chr(0x01):
    raise IOError, "DAC didn't ack bottom bits"
  bp.send_stop_bit()
  bp.resetBP()


def setLMHOptions(device, lp, highgain, ladder, disaux):
  #Assemble the 2 data bytes
  datahigh = 0 | ((disaux & 0x01) << 2) | ((lp & 0x04) >> 2)
  datalow = ((lp & 0x03) << 6) | ((highgain & 0x01) << 4) | (ladder & 0x0F)
  bp = RAW_WIRE(device, 115200)
  print "Current LMH options:"
  readLMHOptions(bp)
  print "-------------------"

  #Send
  #bp.BBmode()
  #bp.enter_rawwire()
  #bp.raw_cfg_pins(PinCfg.CS)
  #if not bp.cfg_raw_wire((RAW_WIRECfg.BIT_ORDER & RAW_WIRE_BIT_ORDER_TYPE.MSB) |
  #                       (RAW_WIRECfg.WIRES & RAW_WIRE_WIRES_TYPE.TWO) |
  #                       (RAW_WIRECfg.OUT_TYPE & RAW_WIRE_OUT_TYPE._3V3)):
  #  raise IOError, "Failed to set pin type correctly"
  #bp.set_speed(RAW_WIRESpeed._400KHZ)
  #bp.CS_Low()
  #bp.bulk_trans(3, [0x00, datahigh, datalow])
  #bp.CS_High()
  #bp.resetBP()

  print "New LMH options:"
  readLMHOptions(bp)
  print "-------------------"

def readLMHOptions(bp):
  bp.BBmode()
  bp.enter_rawwire()
  bp.raw_cfg_pins(PinCfg.CS)
  if not bp.cfg_raw_wire((RAW_WIRECfg.BIT_ORDER & RAW_WIRE_BIT_ORDER_TYPE.MSB) |
                         (RAW_WIRECfg.WIRES & RAW_WIRE_WIRES_TYPE.TWO) |
                         (RAW_WIRECfg.OUT_TYPE & RAW_WIRE_OUT_TYPE._3V3)):
    raise IOError, "Failed to set pin type correctly"
  bp.set_speed(RAW_WIRESpeed._400KHZ)
  bp.CS_Low()
  bp.bulk_trans(1, [0xFF])
  for i in range (0,15):
    print bp.read_bit()
  bp.CS_High()
  bp.resetBP()
  
  print "raw: %u %u" % (datahigh, datalow)

  data = (datahigh << 8) | datalow
  if (data & 0x0200) == 0:
    print "Power: full"
  else:
    print "Power: aux hi-z"

  bw = ["full", "20", "100", "200", "350", "650", "750", "x"]
  value = (data & 0x00E0) >> 6
  print "Filter: %s MHz" % bw[value]

  if (data & 0x0010) == 0:
    print "Pre-amp: low gain"
  else:
    print "Pre-amp: high gain"

  ladder = ["0", "-2", "-4", "-6", "-8", "-10", "-12", "-14", "-16", "-18", "-20", "x", "x", "x", "x", "x"]
  value = data & 0x000F
  print "Attenuation: %s dB" % ladder[value]

#--------------------------------------------------------------------
#Main program routine
#--------------------------------------------------------------------
def main():
  #Parse CLI arguments
  parser = OptionParser()
  parser.add_option("-d", "--device", dest="device", default="/dev/ttyUSB0",
                    help="device name (path or COMx)", metavar="/dev/ttyUSB0")
  parser.add_option("-m", "--mode", dest="mode",
                    help="offset/trig/lmh6518", metavar="offset")
  dacgroup = OptionGroup(parser, "DAC only", "")
  dacgroup.add_option("-v", "--value", dest="value", type="int",
                    help="value for DAC (offset/trig) output", metavar=0x00)
  parser.add_option_group(dacgroup)
  lmhgroup = OptionGroup(parser, "LMH6518 only", "")
  lmhgroup.add_option("-f", "--filter", dest="lp", default=0x00, type="int",
                      help="numeric (3 bit) value to choose filter (Table 6)")
  lmhgroup.add_option("-l", "--ladder", dest="ladder", default=0x00, type="int",
                      help="numeric (4 bit) value to choose ladder (Table 7)")
  lmhgroup.add_option("-H", "--high-gain", action="store_true", dest="highgain",
                      help="Enable to use the high-gain preamp", default=False)
  lmhgroup.add_option("-a", "--dis-aux", dest="disaux", action="store_true",
					  default=False, help="Disable the auxilary output")
  parser.add_option_group(lmhgroup)

  (options, args) = parser.parse_args()

  #Check sanity of received arguments
  lmhmode = ["lmh6518"]
  dacmode = ["offset", "trig"]
  dacaddr = [0x0D, 0x0C]
  modes = lmhmode + dacmode
  if options.mode not in modes:
    parser.error("Mode argument not known")

  if options.mode in dacmode:
    setDACValue(options.device, dacaddr[dacmode.index(options.mode)],
                options.value)
  elif options.mode in lmhmode:
    setLMHOptions(options.device, options.lp, options.highgain, options.ladder, options.disaux)

if __name__ == "__main__":
  main()

File: testboard-custom/testboard.rb
#!/usr/bin/ruby

require 'rubygems'
require 'optparse'
require 'analogdebug'

include Scope

def dp(x)
  x.each { |b|
    print b.to_s(16) + " "
  }
  print "\n"
end


#Define options
options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: testboard.rb [options]"
  opts.separator "Common options:"
  opts.on("-m", "--mode [MODE]", [:offset, :trig, :lmh6518], "Mode (offset/trig/lmh6518)") do |m|
    options[:mode] = m
  end
  opts.separator ""
  opts.separator "DAC only options:"
  opts.on("-v", "--value [VALUE]", OptionParser::DecimalInteger, "10bit value for DAC") do |v|
    options[:value] = v
  end
  opts.separator ""
  opts.separator "LMH6518 only options:"
  opts.on("-f", "--filter [FILTER]", OptionParser::DecimalInteger, "3bit value for filter (Table 6)") do |f|
    options[:filter] = f
  end
  opts.on("-l", "--ladder [LADDER]", OptionParser::DecimalInteger, "4bit value for ladder (Table 7)") do |l|
    options[:ladder] = l
  end
  opts.on("-g", "--[no-]gain", "(No) pre-amp high gain") do |g|
    options[:gain] = g
  end
  opts.on("-a", "--[no-]aux", "(No) auxilary output") do |a|
    options[:aux] = a
  end
end.parse!

p options

dacmode = [:offset, :trig]

#Connect to board
board = AnalogDebug.new

def readDAC(board, addr)
  packet = [0x80, addr]
  board.write packet
  dp packet
  data = board.read.unpack("C*")
  dp data
  if data[0] == 0x81 then
    value = (data[2] << 8) | data[1]
    puts "DAC: #{value}"
  else
    puts "Received wrong packet"
  end
end

#Check option combinations are valid + send
if dacmode.include?(options[:mode]) then
  unless options.include?(:value) then
    puts "When DAC, need value"
    exit 1
  end

  addr = 0x0D if options[:mode] == :offset
  addr = 0x0C if options[:mode] == :trig

  puts "Currently:"
  readDAC(board, addr)

  #Send new
  puts "Sending new:"
  packet = [0x82, addr, options[:value] & 0xFF, (options[:value]>>8) & 0x03]
  dp packet
  board.write packet

  puts "Now:"
  readDAC(board, addr)
else
  unless options.include?(:filter) then
    puts "When LMH6518, need filter"
    exit 1
  end
  unless options.include?(:ladder) then
    puts "When LMH6518, need ladder"
    exit 1
  end
  unless options.include?(:gain) then
    puts "When LMH6518, need gain"
    exit 1
  end
  unless options.include?(:aux) then
    puts "When LMH6518, need aux"
    exit 1
  end

  #Read existing
  puts "Currently:"
  packet = "\x84"
  board.write packet
  dp packet
  data = board.read
  dp data
  if data[0] == "\x85" then
    reg = (data[2] + data[1]).unpack("S").shift
    if reg[10] == 1 then
      puts "Power: auxilary hi-z"
    else
      puts "Power: full"
    end
    filter = ["full", "20", "100", "200", "350", "650", "750", "x"]
    puts "Filter: #{filter[(reg >> 6) & 0x07]}"
    if reg[4] == 1 then
      puts "Pre-amp: high gain"
    else
      puts "Pre-amp: low gain"
    end
    ladder = ["0", "-2", "-4", "-6", "-8", "-10", "-12", "-14", "-16", "-18", "-20", "x", "x", "x", "x", "x"]
    puts "Ladder: #{ladder[reg & 0x0F]}"
  else
    puts "Received wrong packet"
  end


  #Send new
  puts "Sending new:"
  data = 0
  if options[:aux] == 1 then
    data = data | (1 << 10)
  end
  if options[:gain] == 1 then
    data = data | (1 << 4)
  end
  data = data | ((options[:filter] & 0x07) << 6)
  data = data | (options[:ladder] & 0x04)
  packet = "\x86#{(data & 0xFF).chr}#{((data >> 8) & 0xFF).chr}"
  board.write packet
  dp packet
end

